### PR TITLE
NuGet Pack

### PR DESCRIPTION
**Pull request by prabirshrestha**
this is a task for creating NuGet packages - *.nupkg files.

It can be created using:

nuget :create_package do |n|
    n.command = "c:\NuGet\NuGet.exe"
    n.nuspec_file = "D:\Prabir\Desktop\Working\NuPack\Facebook\4.1.1\Facebook.nuspec"
    n.output = "D:\Prabir\Documents\Projects\facebooksdk\temp\Build"
end

only nuspec_file is required, others are optional.

incase n.command is not specified it will set it as "NuGet" by default. (some of the users may have put NuGet.exe in environment path)

nuget :create_package do |n|
    n.nuspec_file = "D:\Prabir\Desktop\Working\NuPack\Facebook\4.1.1\Facebook.nuspec"
    n.output = "D:\Prabir\Documents\Projects\facebooksdk\temp\Build"
end

n.output is also optional. (it default to the same directory where the rakefile.rb is present)

so this is the minmal amount of code that can be written to create a package:
nuget :create_package do |n|
    n.nuspec_file = "D:\Prabir\Desktop\Working\NuPack\Facebook\4.1.1\Facebook.nuspec"
end

another option that may be used is n.base_folder. 

More information about the NuGet command line can be found at http://nuget.codeplex.com/documentation?title=Creating a Package

### CODE REVIEW

**Comment by mxriverlynn** [comment]
i agree w/ steve's suggestions on the .nuspec and defaulting the .command. are these changes already made? or going to be made? 

i'd like to get this pulled in soon. if you don't think you'll have time to make the changes, i can do it after i get the changes pulled into my repo.

**Comment by prabirshrestha** [comment]
thanks for the reminder. was about it do it later when i totally forgot about it.

i changed the default parameters in initialize method and also changed nuspec_file to nuspec.

**Comment by prabirshrestha** [comment]
and recently nuget also supported the push command which i havent implemented yet.
the current nuget task implements pack only.

Should we have two tasks for nuget or just one task for nuget?

The command line to create nuget package, 
usage: NuGet pack <manifest> [options]

to push
usage: NuGet push <package location> <API key> [options]

**Comment by stevehebert** [comment]
I created a separate pull request on a nuget task that validates the required parameters, respects the optional parameters, generates the nuspec file and runs nuget pack against it.  This gives you the opportunity to drive the version number in the file from rake which I really like.

I like the idea of a separate push option.  My oss builds look at push as a non-automatic task, but something that sits in TeamCity ready to go.

**Comment by prabirshrestha** [comment]
saw your pull request too. looks great.
but not sure if there should be a different task to generate the nuspec file or combine it as the same task.
or we could have a option like yours generate the nuspec file on the fly but if it has a parameter called, nuspec then it rather looks at the file.
or another option is to load from nuspec file if present, and the parameter overrides the file.

**Comment by stevehebert** [comment]
I believe the task of generating and packaging should be together.

If you look at nuget repo behavior, once a version of the package is uploaded once it cannot be uploaded again.  A given version of a package is considered immutable in nuget repositories. Therefore, a new build of the software should have a unique version associated if it is be useful to nuget and building against a static nuget file has limited value at least from what I've seen.

I think the package upload process should be a separate step because not all builds will want to perform the upload.  This is especially the case when hosting your own nuget server - where your push process is slightly different.

**Comment by prabirshrestha** [comment]
For uploading we would definitely need a different task for it.

For generating and packing might be we could do together too.

**Comment by mxriverlynn** [comment]
my understanding of nuget and how it works is fairly limited, as i've only seen it demo'd a few times... but i know it's based on rubygems and how that system works, so i think i have an abstract understanding of it. given all that, here's how i see things being set up:
- .nuspec file generating
- nuget package creation
- nuget package deployment (push to whatever server you're using)

for the .nuspec file generating, i can see 2 ways of handling this: a "nuspec" task - take the work that steve hebert did and split it into a "nuspec" task - or simple template systems like ERB or HAML. i like the idea of a nuspec task to generate everything, so that it's easily configured and easy to see what's going on directly from the rake task.

for the nuget package creation - i think we all agree on how simple this can be. both steve hebert and prabir have tasks for this. it may be easier to use prabir's in this case, since we would not have to take anything out of the task.

for the push / deploy process - i'm a little less clear on this. there's a way to have nuget push a package up to the standard package servers, right? what servers are there? is nuget.org the official server? or is there somethign on codeplex? or ??? ... then there's the need for internal package servers. i know several people that use internal servers for various packages already. we probably don't need to handle the internal server deploys with an albacore task directly, but i think we should have a "nugetpush" or "nugetdeploy" task to push out to nuget.org (or whatever the official server is).

so, to summarize, i think we should:
- use steve hebert's code as a "nuspec" task, to generate the nuspec file
- use prabir's code as a "nuget" or "nugetpackage" task, to build the nuget packages
- use ??? code as a "nugetdeploy" or "nugetpush" task, to push built packages out to nuget.org

what do you guys think? am i on the right track, or am i off in crazy land again? :)

**Comment by prabirshrestha** [comment]
your on the right track.

for me i think it is better to separate the nuspec and nuget (pack) task. This allows the user to be in more control.

And as for the push/deploy there are things that can be configured. (http://nuget.org/Contribute/Index) codeplex one has been obsolete and all new packages should be pushed to www.nuget.org
here is the copy paste help command for "nuget help push".
usage: NuGet push <package location> <API key> [options]

Push a package to the server and publish it to the feed.

```
 Specify the location of the package and your API key to push the package to
 the server.
```

options:

 Publish   (pub)    Specifies if the package should be published to the live fee
                    d. True by default.
 Source    (src)    Specifies the base GalleryServer Url.

Hopefully this answers the third task nugetdeploy or nugetpush.

**Comment by stevehebert** [comment]
I see your point - support the three tasks discretely and let the user build them up.  It's easy to change mine to remove the execution behavior.

I'm researching how to best include the dependencies and files collections with nuget and then we're good to go.  My ruby skills are pretty limited at this point, so I am looking at other Albacore tasks as guideposts.

**Comment by mxriverlynn** [comment]
ok, sounds like we should call the push task "nugetpush", and may want to called the packaging task "nugetpack", to reflect the command line that is called for each of these.

**Comment by mxriverlynn** [comment]
let me know if  you want any help with the ruby side of things, steve. or feel free to ask questions on the google group, to get a wider audience: http://groups.google.com/group/albacoredev/

**Comment by prabirshrestha** [comment]
done for NuGetPack.
Here is a sample of all the available parameters available.

```
nugetpack :pack do |n|
    n.command = "tools/nuget/nuget.exe"
    n.nuspec = "src/Facebook.nuspec"
    n.output = "dist/"
    n.base_folder = "bin/"
end
```

command is optional. if not specified it defaults to "NuGet.exe" as some may have put it in the environement path.

nuspec is required.
if not specified it will show an error: FATAL -- : nuspec must be specified.

output and base_folder are optional. (as it is optional for nuget.exe)
here is the description of it from "nuget help pack"
 OutputDirectory   (o)    Specifies the directory for the created NuGet file

 BasePath          (b)    The Base Path of the files defined in the nuspec file.

**Comment by prabirshrestha** [comment]
and i renamed "nuget" to "nugetpack" so later on we can have the "nugetpush" task

**Comment by stevehebert** [comment]
and I've renamed my "nuget" to "nuspec".  I sent a code question to Derick and I'll submit a pull request when I have the dependencies and file collections working.

**Comment by stevehebert** [comment]
Prepping my pull request now - the nuspec task will look like this in the rake file:

<pre><code>
nuspec do |nuspec|
   nuspec.id="fluentworkflow"
   nuspec.output_file = "fluentworkflow.nuspec"
   nuspec.version = full_version
   nuspec.authors = "Steve Hebert"
   nuspec.description = "FluentWorkflow is a workflow/saga dsl and runtime component"
   nuspec.language = "en-US"
   nuspec.licenseUrl = "http://fluentworkflow.codeplex.com/license"
   nuspec.projectUrl = "http://fluentworkflow.codeplex.com"
   nuspec.working_directory = "Build/Deploy"
   nuspec.dependency "Autofac", "2.3.2.632"
   nuspec.dependency "Stateless"
   nuspec.file "bin/Debug/*.dll", "lib"
   nuspec.file "bin/Debug/*.pdb", "lib"
end
</code></pre>


and will produce xml that looks like this:

https://gist.github.com/789019

**Comment by stevehebert** [comment]
I've updated the pull request for the 'nuspec' task.

**Comment by mxriverlynn** [comment]
pulled prabir's code into my local branch. am working on getting other requests pulled in, too.
